js/runtime.go

```diff
@@ -123,22 +123,44 @@ func (r *Runtime) extractOptions(exports otto.Value, opts *lib.Options) error {
 	return nil
 }
 
-func (r *Runtime) loadFile(filename string) (otto.Value, error) {
-	// To protect against directory traversal, prevent loading of files outside the root (pwd) dir
-	path, err := filepath.Abs(filename)
+func (r *Runtime) loadJSCode(src *lib.SourceData) (otto.Value, error) {
+	var path string
+	var err error
+	path, err = filepath.Abs(src.Filename)
+	if err != nil {
+		return otto.UndefinedValue(), err
+	}
+
+	// Don't re-compile repeated includes of the same module
+	if exports, ok := r.Exports[path]; ok {
+		return exports, nil
+	}
+	exports, err := r.load(path, src.SrcData)
 	if err != nil {
 		return otto.UndefinedValue(), err
 	}
-	if !strings.HasPrefix(path, r.Root) {
-		return otto.UndefinedValue(), DirectoryTraversalError{Filename: filename, Root: r.Root}
+	r.Exports[path] = exports
+
+	log.WithField("path", path).Debug("File loaded")
+
+	return exports, nil
+}
+
+func (r *Runtime) loadFile(filename string) (otto.Value, error) {
+	var path string
+	var data []byte
+	var err error
```

Comment:
This preamble is giving me C89 vibes. You can accomplish the same thing with some `:=`'s.